src/app/clusters/mode-base-server/mode-base-cluster-objects.h

```diff
@@ -0,0 +1,253 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/util/util.h>
+#include <utility>
+
+#define EMBER_AF_PLUGIN_MODE_BASE
```

Comment:
The public interface should NOT be dependent on these defines. Only use these in implementations. These EMBER_AF defines are very specific to code gen and should not appear in common code.

Reply:
This is currently required because of the entanglement of the ModeBase cluster with the OnOff cluster. We can get rid of this once the OnOff cluster provides an API to interact with it. 

Also, because the ModeBase cluster is a primary cluster for other clusters to be derived from, zap will not define this in `zapgen/zap-generated/gen_config.h` as this cluster is never used. Something else to be added to zap for aliased/derived clusters.

Reply:
Actually, this may not be the correct location for it. It stopped working. I think it's since I fixed the imports for the OnOff server code. Now the OnOff server is not picking up this define.

My logic to add it here was that if an SDK consumer is going to have a ModeBase derived cluster, then they will need to include this file in their application.

Reply:
It still does not work because the OnOff server is not picking up the definition. 

Should I omit the use of it and write a function that checks if any of the ModeBase clusters are present on the endpoint?

Reply:
I think the right things to do are:

1. Have some sort of config variable that says "Mode Base is enabled".
2. Condition the On/Off code on that config variable.
3. Have the mode base .cpp compile fail if that variable is not defined true (via `#error`).

What that variable is called... doesn't matter that much in the end.  You could call it CHIP_CONFIG_MODE_BASE_ENABLED, or you could have something like that which, if defined, leads to `EMBER_AF_PLUGIN_MODE_BASE` being defined and use that....

Reply:
Is [this](https://github.com/project-chip/connectedhomeip/pull/27504/commits/995bdd18a604bc48d61784d2b95a3146796687e8) a good solution? 

I don't know which location in the SDK would be possible to add a global variable that can be accessed by both the ModeBase and OnOff clusters irrespective of either of them being used.

Reply:
For now the solution will do. I think what @bzbarsky-apple was alluding to was BUILD.gn changes.

Reply:
I have filed [this](https://github.com/project-chip/connectedhomeip/issues/28155) issue to keep track of making this change.

Reply:
I am pretty agnostic on where the config variable comes from.... Not least because our config situation is so broken, I don't even know where to start on what should go where config-wise.